nightfall-deployer/contracts/Challenges.sol

```diff
@@ -268,13 +268,33 @@ contract Challenges is Stateful, Config {
             transaction.transactionSiblingPath
         );
 
-        for (uint256 i = 0; i < transaction.transaction.nullifiers.length; ++i) {
-            uint64 historicblockNumberL2 = Utils.getHistoricRoot(
+        // Unpack the historic block numbers, they are packed in 256bits and so 4 x 64 historicRootBlockNumbers are unpacked.
+        uint64[] memory historicRootBlockNumbers = new uint64[](
+            transaction.transaction.historicRootBlockNumberL2.length * 4
+        );
+
+        for (uint256 i = 0; i < historicRootBlockNumbers.length; ++i) {
+            historicRootBlockNumbers[i] = Utils.getHistoricRoot(
                 transaction.transaction.historicRootBlockNumberL2,
                 i
             );
+        }
+
+        for (uint256 i = 0; i < transaction.transaction.nullifiers.length; ++i) {
+            if (uint256(historicRootBlockNumbers[i]) >= state.getNumberOfL2Blocks()) {
+                challengeAccepted(transaction.blockL2);
+                return;
+            }
+        }
 
-            if (uint256(historicblockNumberL2) >= state.getNumberOfL2Blocks()) {
+        // Historic Block Numbers are packed in 256bits and so always results in multiples of 4
+        // if nullifiers are not multiple of 4 we need to ensure these "extra" historic block numbers are zero.
+        for (
+            uint256 i = historicRootBlockNumbers.length;
+            i < transaction.transaction.nullifiers.length;
```

Comment:
I think this is probably wrong. I would expect `i > transaction.transaction.nullifiers.length`.  So `>` instead of `<`.

Reply:
Good catch! Yep it should be the other way - I'll push up a fix

Reply:
I think I also need to be `i = historicRootBlockNumbers.length -1` in the first line